node/reqres.js

```diff
@@ -105,12 +124,26 @@ TChannelOutgoingRequest.prototype.hookupCallback = function hookupCallback(callb
     self.once('error', onError);
     self.once('response', onResponse);
     function onError(err) {
+        // TODO: better annotations
+        self.span.annotate('cr'); // client recv
+        self.tracer.report(self.span);
         self.removeListener('response', onResponse);
+
+        var prev = self.tracer.getCurrentSpan();
+        self.tracer.setCurrentSpan(self.span);
         callback(err, null);
+        self.tracer.setCurrentSpan(prev);
     }
     function onResponse(res) {
+        // TODO: better annotations
+        self.span.annotate('cr');
+        self.tracer.report(self.span);
         self.removeListener('error', onError);
+
+        var prev = self.tracer.getCurrentSpan();
+        self.tracer.setCurrentSpan(self.span);
         callback(null, res);
+        self.tracer.setCurrentSpan(prev);
```

Comment:
`async-listener` takes care of setting and restoring the context during any async operations. Note that the tests use `setTimeout` and `process.nextTick` to test this.